x-pack/plugins/secret_service/server/service/secret_service.ts

```diff
@@ -0,0 +1,125 @@
+/*
+ * Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+ * or more contributor license agreements. Licensed under the Elastic License;
+ * you may not use this file except in compliance with the Elastic License.
+ */
+
+// @ts-ignores
+import nodeCrypto from '@elastic/node-crypto';
+import crypto from 'crypto';
+import { SavedObjectsClient } from 'src/legacy/server/saved_objects';
+// @ts-ignore
+import { AuditLogger } from '../../../server/lib/audit_logger';
+
+type SSEvents = 'secret_object_created' | 'secret_object_accessed' | 'secret_object_decrypt_failed';
+type SecretId = Promise<string>;
+
+export class SecretService {
+  public readonly hideAttribute: (deets: any) => SecretId;
+  public readonly unhideAttribute: (deets: any) => any;
+  public readonly validateKey: () => Promise<boolean>;
+  private readonly type: string;
+  private readonly auditor?: AuditLogger;
+  private readonly hideAttributeWithId: (id: string, toEncrypt: any) => SecretId;
+
+  constructor(
+    savedObjectsClient: SavedObjectsClient,
+    type: string,
+    key?: Buffer,
+    auditor?: AuditLogger
+  ) {
+    this.type = type;
+    this.auditor = auditor;
+    key = key || crypto.randomBytes(128);
+    const crypt = nodeCrypto({ encryptionKey: key.toString('hex') });
+
+    const logEvent = (event: SSEvents, message?: string) => {
```

Comment:
If they were private instance methods all the details including the encryption key would be exposed. I think it's a fine tradeoff to make these declared anonymously to reduce the risk of being caught by other unforeseen vulnerabilities. 

We can create the methods for docs and call internal methods but the purpose of that seems less clear and mostly unnecessary from the implementation and easily forces someone to desire to refactor it. Not saying doing it this way doesn't also push for the same refactoring desire.

`crypt` shouldn't be an instance property for the same reason the hide, unhide, etc weren't declared as instance methods. But since we are taking the stance that we trust all the code running in kibana and if the keystore is accessible to all then I don't know that my concerns bare much weight other than to say I'm uncomfortable writing code that leaves itself vulnerable to known attack vectors without multiple layers of security.

Reply:
I see, it still feels like we're trading readability for a questionable increase in security since, as you mentioned, every piece of Kibana code can access full config and file system anyway. At some point TS `private` will become "real" private and there is a chance that you won't expose entire service to the consumers anyway as it may become more complex in the future (`service.validateKey` is already something that ideally should not be a part of public API exposed to the plugins). But it's just bike-shedding, so up to you.